contracts/ConversionRates2.sol

```diff
@@ -0,0 +1,426 @@
+        if (command == 11) return tokenData[token].buyRateImbalanceStepFunction.y[param];
+
+        if (command == 12) return int(tokenData[token].sellRateImbalanceStepFunction.x.length);
+        if (command == 13) return tokenData[token].sellRateImbalanceStepFunction.x[param];
+        if (command == 14) return int(tokenData[token].sellRateImbalanceStepFunction.y.length);
+        if (command == 15) return tokenData[token].sellRateImbalanceStepFunction.y[param];
+
+        revert();
+    }
+    /* solhint-enable code-complexity */
+
+    function getRateUpdateBlock(ERC20 token) public view returns(uint) {
+        bytes32 compactData = tokenRatesCompactData[tokenData[token].compactDataArrayIndex];
+        return getLast4Bytes(compactData);
+    }
+
+    function getListedTokens() public view returns(ERC20[]) {
+        return listedTokens;
+    }
+
+    function getTokenQty(ERC20 token, uint ethQty, uint rate) internal view returns(uint) {
+        uint dstDecimals = getDecimals(token);
+        uint srcDecimals = ETH_DECIMALS;
+
+        return calcDstQty(ethQty, srcDecimals, dstDecimals, rate);
+    }
+
+    function getLast4Bytes(bytes32 b) internal pure returns(uint) {
+        // cannot trust compiler with not turning bit operations into EXP opcode
+        return uint(b) / (BYTES_14_OFFSET * BYTES_14_OFFSET);
+    }
+
+    function getRateByteFromCompactData(bytes32 data, ERC20 token, bool buy) internal view returns(int8) {
+        uint fieldOffset = tokenData[token].compactDataFieldIndex;
+        uint byteOffset;
+        if (buy)
+            byteOffset = 32 - NUM_TOKENS_IN_COMPACT_DATA + fieldOffset;
+        else
+            byteOffset = 4 + fieldOffset;
+
+        return int8(data[byteOffset]);
+    }
+
+    function executeStepFunction(StepFunction f, int x) internal pure returns(int) {
+        if (x == 0) { return 0; }
+        uint len = f.y.length;
+        int lastStepAmount = 0; // amount from last step to compute amount to be applied bps for next step
+        int change = 0; // amount change from initial amount when applying bps for each step
+        uint ind;
+        if (x > 0) {
+            for (ind = 0; ind < len; ind++) {
+                if (f.x[ind] <= 0) { continue; } // ignore non-positive steps
+                if (x <= f.x[ind]) {
+                    change += (x - lastStepAmount) * f.y[ind];
+                    lastStepAmount = x;
+                    break;
+                }
+                change += (f.x[ind] - lastStepAmount) * f.y[ind];
+                lastStepAmount = f.x[ind];
+            }
+            if (x > lastStepAmount) {
+                change += (x - lastStepAmount) * f.y[len - 1];
+            }
+            return change / x;
+        }
+        // revert steps order for negative number
+        int lastStepBps = 0;
+        uint id;
+        for (ind = len; ind > 0; ind--) {
+            id = ind - 1;
```

Comment:
not sure why you need both ind and id
seems like ind isn't used at all in the loop